lib/module-entry.js

```diff
@@ -0,0 +1,227 @@
+  this._sourceHash = sourceHash;
+
+  // the computed hash for the package's source files plus
+  // the source hashes of the dependencies. Because we compute
+  // dependencies as 'all the way down', this gets everything.
+  this._hash = null;
+
+  // references to other ModuleEntry objects that correspond
+  // to the dependencies listed in the package.json
+  this._dependencies = null;
+}
+
+ModuleEntry.prototype.initializeDependencies = function(caches, pkg, hashTreeFn) {
+  if (this._dependencies !== null) {
+    throw new Error('[hashForDep] Cannot re-initializeDependencies a moduleEntry');
+  }
+  this._dependencies = Object.create(null);
+
+  if (!pkg.dependencies) { return; }
+
+  var moduleEntry = this;
+  var Constructor = this.constructor;
+  // Recursively locate the references to other moduleEntry for the package's
+  // dependencies. Initially we just do the 'local' dependencies (i.e. the ones referenced
+  // in the package.dependencies field.) Later, hashForDep calls 'gatherDependencies'
+  // to compute the complete list including transitive dependencies, and that is then
+  // used for the final hash calculation.
+  Object.keys(pkg.dependencies).sort().forEach(function(dep) {
+    var dependencyModuleEntry = Constructor.locate(caches, dep, moduleEntry.rootDir, hashTreeFn);
+    // there's not really a good reason to include a failed resolution
+    // of a package in the dependencies list.
+    if (dependencyModuleEntry !== null) {
+      moduleEntry._dependencies[dep] = dependencyModuleEntry;
+    }
+  });
+};
+
+/*
+ * Compute the 'hash' field for the given moduleEntry.
+ * The hash actually consists only of the sourceHash values of all the
+ * dependencies (PackageCacheEntry objects) concatenated with '|' between
+ * entries. The 'sourceHash' values are computed at the time the
+ * PackageCacheEntry is created, so don't need to be done again.
+ */
+ModuleEntry.prototype.getHash = function(heimdallNode) {
+  if (this._hash !== null) {
+    return this._hash;
+  }
+
+  // We haven't computed the hash for this entry yet, though the
+  // packageCache entries are all present. It is possible that we
+  // have the hash for some dependencies, though.
+
+  // compute the full dependency list. There are no duplicates, because
+  // we check during gatherDependencies before inserting in the result.
+  // The keys of the returned object are the package baseDir values (the
+  // moduleEntry.rootDir). The values are the moduleEntry
+  // objects.
+  var dependencies = this._gatherDependencies(Object.create(null));
+
+  // For repeatability between runs, sort the dependencies list keys
+  // (rootDir values) before assembling into a final hash value.
+  var dependencyRootDirs = Object.keys(dependencies).sort();
+  heimdallNode.stats.paths += dependencyRootDirs.length;
+
+  var sourceHashes = dependencyRootDirs.map(function(rootDir) {
+    return dependencies[rootDir]._sourceHash;
+  }).join('|');
```

Comment:
should we use `"\x00"` instead of '|' ? So we have a null char, and no accidental collisions?

Reply:
See earlier reply - 0x00 doesn't do what you think it does when + into a string.

Reply:
I believe it does `"\x00".charCodeAt(0) === 0` 

I believe i simply made a typo earlier and did `0x00` instead of `\x00`

Reply:
Ah, got it. Let me test, and if it's working, I'm fine with using it. It did always seem kind of odd. :-)

Reply:
I switched to using \x00 except in hash-tree (I think you were using that already, right?) If you want it there, too, fine with me.